mist-core/src/main/java/edu/snu/mist/core/sources/MQTTSubscribeClient.java

```diff
@@ -84,27 +96,90 @@ public MQTTDataGenerator connectToTopic(final String topic) {
     Queue<MQTTDataGenerator> dataGeneratorList = dataGeneratorListMap.get(topic);
     if (dataGeneratorList == null) {
       dataGeneratorList = new ConcurrentLinkedQueue<>();
-      dataGeneratorListMap.put(topic, dataGeneratorList);
+      dataGeneratorListMap.putIfAbsent(topic, dataGeneratorList);
     }
     final MQTTDataGenerator dataGenerator = new MQTTDataGenerator(this, topic);
     dataGeneratorListMap.get(topic).add(dataGenerator);
     return dataGenerator;
   }
 
   /**
-   * Start to subscribe a topic.
+   * Connect to client.
    */
-  void subscribe(final String topic) throws MqttException {
-    synchronized (subscribeLock) {
-      if (!started) {
+  private void connect() {
+    while (true) {
+      try {
         client = new MqttAsyncClient(brokerURI, clientId);
         final MqttConnectOptions mqttConnectOptions = new MqttConnectOptions();
         mqttConnectOptions.setKeepAliveInterval(mqttSourceKeepAliveSec);
         client.connect(mqttConnectOptions).waitForCompletion();
         client.setCallback(this);
+        break;
+      } catch (final MqttException e) {
+        try {
+          client.close();
+        } catch (final Exception e1) {
+          // do nothing
+        }
+        // Reconnect mqtt
+        LOG.log(Level.SEVERE, "Connection for broker {0} with id {1} failed ... Retry connection",
+            new Object[] {brokerURI, clientId});
+        try {
+          Thread.sleep(1000);
+        } catch (final InterruptedException e1) {
+          e1.printStackTrace();
+        }
+      }
+    }
+  }
+
+  /**
+   * Start to subscribe a topic.
+   */
+  void subscribe(final String topic) {
```

Comment:
Why no `private` here?

Reply:
This was not private method because it is used `MqttDataGenerator`